Chatkit/Buffer/Buffer.swift

```diff
@@ -0,0 +1,130 @@
+
+class ConcreteBuffer: Buffer {
+    
+    // MARK: - Types
+    
+    typealias Dependencies = HasStore
+    
+    // MARK: - Properties
+    
+    private(set) var currentState: VersionedState?
+    private var queue: [VersionedState]
+    
+    private let dependencies: Dependencies
+    
+    let filter: StateFilter
+    
+    weak var delegate: BufferDelegate?
+    
+    // MARK: - Initializers
+    
+    init(filter: StateFilter, dependencies: Dependencies) {
+        self.currentState = nil
+        self.queue = []
+        self.filter = filter
+        self.dependencies = dependencies
+        
+        self.registerListener()
+    }
+    
+    // MARK: - Store listener
+    
+    func store(_ store: Store, didUpdateState state: VersionedState) {
+        self.apply(state: state)
+    }
+    
+    // MARK: - Private methods
+    
+    private func registerListener() {
+        // It really feels like this should be done by a different object, but such approach is forced on this class by the DI mechanism.
+        let state = self.dependencies.store.register(self)
+        
+        if self.filter.hasCompleteSubstate(state) {
+            self.currentState = state
+        }
+        else {
+            // The current state of the store cannot be reported until it gets supplemented by one of the subsequent states.
+            self.enqueue(state: state)
+        }
+    }
+    
+    private func apply(state: VersionedState) {
+        self.enqueueIfNeeded(state: state)
+        self.flush(withSupplementalState: state)
+    }
+    
+    private func enqueueIfNeeded(state: VersionedState) {
+        if let currentState = self.currentState {
+            let lastState = self.queue.last ?? currentState
+            
+            if self.filter.hasSupportedSignature(state.signature)
+                && self.filter.hasModifiedSubstate(oldState: lastState, newState: state) {
+                self.enqueue(state: state)
+            }
+        }
+        else if self.filter.hasSupportedSignature(state.signature) {
+            self.enqueue(state: state)
+        }
+    }
+    
+    private func enqueue(state: VersionedState) {
+        self.queue.append(state)
+    }
+    
+    private func flush(withSupplementalState supplementalState: VersionedState) {
+        var removedIndexes: [Int] = []
+        
+        for (index, state) in self.queue.enumerated() {
+            if self.filter.hasCompleteSubstate(state) {
+                self.currentState = state
+                self.delegate?.buffer(self, didUpdateState: state)
+                removedIndexes.append(index)
+            }
+            else {
+                let supplementedState = state.supplement(withState: supplementalState)
+                
+                if self.filter.hasCompleteSubstate(supplementedState) {
+                    self.currentState = supplementedState
+                    self.delegate?.buffer(self, didUpdateState: supplementedState)
```

Comment:
This would mean that `registerListener()` method needs to be changed. Otherwise, we would call the delegate method when initializing with a complete substate.

Reply:
I don't think it does.  The setting of `state` in the `registerListener` will *not* invoke the delegate because `didSet`'s are not invoke during initialisation (see below).  If you make the change your unit tests will still continue to pass.

From https://docs.swift.org/swift-book/LanguageGuide/Properties.html
> Note
> You must declare computed properties—including read-only computed properties—as variable properties with the var keyword, because their value is not fixed. The let keyword is only used for constant properties, to indicate that their values cannot be changed once they are set as part of instance initialization.

Reply:
I am aware about this comment in the documentation. However, it appears that the documentation is not very precise here. Please feel free to run the following code to test the actual behaviour:

```swift
class Foo {
    
    var value: String {
        didSet {
            print("didSet: \(value)")
        }
    }
    
    init() {
        self.value = "First value"
        self.value = "Second value"
        self.modify()
    }
    
    func modify() {
        self.value = "Third value"
    }
    
}

let foo = Foo()
print("\(foo.value)")
```

This should produce the following output: 

```
didSet: Third value
Third value
```

It looks like `didSet` is not being called for all assignments made directly in the initializer, but it is called for all assignments made in methods called from within the initializer.

Reply:
Your example does indeed demonstrate the potential pitfalls of this Swift mechanism, I've learned something 👍🏼  

It seems the Swift compiler is aware of the exact point at which all its stored properties become initialised (and in the example you provided that occurs exactly after `self.value = "First value"`)  From that point onwards `didSet` fires if a given assignment was made in a sub-routine but does not fire if assiged directly in the `init`. 

I'm happy to relinquish my original comment.  However I would like to point out that, at present, its impossible to set the delegate before the the initialiser has returned so IMO there's no real problem anyway.  🤷🏼‍♂️